compiler/src/iree/compiler/Codegen/LLVMGPU/TransformExtensions/LLVMGPUExtensions.cpp

```diff
@@ -815,5 +819,533 @@ DiagnosedSilenceableFailure transform_dialect::ReorderTransposeOp::applyToOne(
+  if (op->hasAttr("__parallel_region_boundary_for_test")) return true;
+
+  // We consider functions inside executable variants that have the same symbol
+  // name as an export symbol.
+  auto func = dyn_cast<FunctionOpInterface>(op);
+  if (!func) return false;
+  auto parent = op->getParentOfType<ModuleOp>();
+  if (!parent) return false;
+  auto variant = parent->getParentOfType<HAL::ExecutableVariantOp>();
+  if (!variant) return false;
+  WalkResult result = variant.walk([&](HAL::ExecutableExportOp exportOp) {
+    if (exportOp.getSymNameAttr() == func.getNameAttr())
+      return WalkResult::interrupt();
+    return WalkResult::skip();
+  });
+  return result.wasInterrupted();
+}
+
+/// Returns `true` if the op behaves like a sequential loop, e.g., the control
+/// flow "wraps around" from the end of the body region back to its start.
+static bool isSequentialLoopLike(Operation *op) { return isa<scf::ForOp>(op); }
+
+/// Returns `true` if the regions of the op are guaranteed to be executed at
+/// most once. Thus, if an operation in one of the nested regions of `op` is
+/// executed than so are all the other operations in this region.
+static bool hasSingleExecutionBody(Operation *op) {
+  return isa<scf::IfOp, memref::AllocaScopeOp>(op);
+}
+
+/// Returns `true` if the operation is known to produce a pointer-like object
+/// distinct from any other object produced by a similar operation. For example,
+/// an allocation produces such an object.
+static bool producesDistinctBase(Operation *op) {
+  return isa_and_nonnull<memref::AllocOp, memref::AllocaOp>(op);
+}
+
+/// Populates `effects` with all memory effects without associating them to a
+/// specific value.
+static void addAllValuelessEffects(
+    SmallVectorImpl<MemoryEffects::EffectInstance> &effects) {
+  effects.emplace_back(MemoryEffects::Effect::get<MemoryEffects::Read>());
+  effects.emplace_back(MemoryEffects::Effect::get<MemoryEffects::Write>());
+  effects.emplace_back(MemoryEffects::Effect::get<MemoryEffects::Allocate>());
+  effects.emplace_back(MemoryEffects::Effect::get<MemoryEffects::Free>());
+}
+
+/// Collect the memory effects of the given op in 'effects'. Returns 'true' if
+/// it could extract the effect information from the op, otherwise returns
+/// 'false' and conservatively populates the list with all possible effects
+/// associated with no particular value or symbol.
+static bool collectEffects(
+    Operation *op, SmallVectorImpl<MemoryEffects::EffectInstance> &effects,
+    bool ignoreBarriers = true) {
+  // Skip over barriers to avoid infinite recursion (those barriers would ask
+  // this barrier again).
+  if (ignoreBarriers && isa<gpu::BarrierOp>(op)) return true;
+
+  // Skip over ops that we know have no effects.
+  if (isKnownNoEffectsOpWithoutInterface(op)) return true;
+
+  // Collect effect instances the operation. Note that the implementation of
+  // getEffects erases all effect instances that have the type other than the
```

Comment:
ugh .. can you comment on that upstream behavior ?
Should it be called takeEffects instead ?
No need to change anything here, I am just curious.

Reply:
Presumably it was just easier to write as `erase_if` with `isa` inside: https://github.com/llvm/llvm-project/blob/d37d4072f2eef35ba573350faa07909a36988e34/mlir/include/mlir/Interfaces/SideEffectInterfaceBase.td#L63-L70